hip-service/Link/Patient/LinkPatient.cs

```diff
@@ -0,0 +1,137 @@
+            this.patientRepository = patientRepository;
+            this.patientVerification = patientVerification;
+        }
+
+        public async Task<Tuple<PatientLinkReferenceResponse, ErrorResponse>> LinkPatients(PatientLinkReferenceRequest request)
+        {
+            var patients = patientRepository.GetPatientInfoWithReferenceNumber(request.Patient.ReferenceNumber);
+
+            return await patients.Map(async (patient) =>
+            {
+                var requestPatient = request.Patient;
+
+                var programs = (from careContext in requestPatient.CareContexts
+                    where patientRepository.GetProgramInfo(requestPatient.ReferenceNumber, careContext.ReferenceNumber)
+                        .HasValue
+                    select patientRepository.GetProgramInfo(requestPatient.ReferenceNumber, careContext.ReferenceNumber)
+                    into careContextPatient
+                    select careContextPatient.Map<CareContext>(context =>
+                    {
+                        var newCareContext = new CareContext
+                            {ReferenceNumber = context.ReferenceNumber, Description = context.Description};
+                        return newCareContext;
+                    })).ToList();
+
+                if (programs.Count != requestPatient.CareContexts.Count())
+                {
+                    return new Tuple<PatientLinkReferenceResponse, ErrorResponse>
+                    (null,
+                        new ErrorResponse(new Error(ErrorCode.CareContextNotFound,
+                            "Care context not found for given patient")));
+                }
+
+                var linkRefNumber = Guid.NewGuid().ToString();
+
+                // method call for generating the OTP
+                var session = new Session(linkRefNumber,
+                    new Communication(CommunicationMode.MOBILE, patient.PhoneNumber));
+
+                if (await patientVerification.SendTokenFor(session) != null)
+                {
+                    return new Tuple<PatientLinkReferenceResponse, ErrorResponse>
+                        (null, new ErrorResponse(new Error(ErrorCode.OtpGenerationFailed, "Unable to create token")));
+                }
+
+                var (_, exception) = await linkPatientRepository.SaveLinkPatientDetails(linkRefNumber,
+                    requestPatient.ConsentManagerId,
+                    requestPatient.ConsentManagerUserId, requestPatient.ReferenceNumber, requestPatient.CareContexts
+                        .Select(context => context.ReferenceNumber).ToArray());
+
+                if (exception != null)
+                {
+                    return new Tuple<PatientLinkReferenceResponse, ErrorResponse>
+                    (null,
+                        new ErrorResponse(new Error(ErrorCode.ServerInternalError,
+                            "Unable to store data to Database")));
+                }
```

Comment:
Can we create some meaningful private methods instead of having big method?

Reply:
Done